crates/circuit/src/parameterexpression.rs

```
@@ -0,0 +1,779 @@
+    }
+    /// return conjugate of expression
+    pub fn conjugate(&self) -> Self {
+        Self {
+            expr: self.expr.conjugate(),
+        }
+    }
+    /// return derivative of this expression for param
+    pub fn derivative(&self, param: &Self) -> Self {
+        Self {
+            expr: self.expr.derivative(&param.expr),
+        }
+    }
+
+    /// expand expression
+    pub fn expand(&self) -> Self {
+        Self {
+            expr: self.expr.expand(),
+        }
+    }
+
+    /// get hashset of all the symbols used in this expression
+    pub fn symbols(&self) -> HashSet<String> {
+        self.expr.symbols()
+    }
+
+    /// return all values in this equation
+    pub fn values(&self, py: Python) -> PyResult<Vec<PyObject>> {
+        let ret: Vec<PyObject> = self
+            .expr
+            .values()
+            .iter()
+            .map(|val| match val {
+                Value::Real(r) => r.into_py_any(py).unwrap(),
+                Value::Int(i) => i.into_py_any(py).unwrap(),
+                Value::Complex(c) => c.into_py_any(py).unwrap(),
+            })
+            .collect();
+        Ok(ret)
+    }
+
+    /// return expression as a string
+    #[getter]
+    pub fn name(&self) -> String {
+        self.__str__()
+    }
+
+    /// bind values to symbols given by input hashmap
+    pub fn bind(&self, py: Python, map: HashMap<String, PyObject>) -> PyResult<Self> {
+        let map: HashMap<String, Value> = map
+            .into_iter()
+            .map(|(key, val)| {
+                if let Ok(i) = val.extract::<i64>(py) {
+                    (key, Value::from(i))
+                } else if let Ok(r) = val.extract::<f64>(py) {
+                    (key, Value::from(r))
+                } else if let Ok(c) = val.extract::<Complex64>(py) {
+                    (key, Value::from(c))
+                } else {
+                    // if unsupported data type, insert empty
+                    ("".to_string(), Value::Int(0))
```

Is this existing behavior, or should we raise an error?